File: nim/bench_ringbuffer.nim
# bench_ringbuffer.nim
# Cross-process SPSC ring buffer over POSIX shared memory, same design
# as shared_memory_ipc.py and bench_ringbuffer.cpp (doubled-modulus
# indices), benchmarked the same way: fork a producer, parent consumes,
# time until all messages are received.

import std/[posix, atomics, monotimes, times, strformat]

const Capacity = 1 shl 20  # 1 MiB

type
  RingHeader = object
    writeIdx: Atomic[uint64]
    readIdx: Atomic[uint64]
    capacity: uint64
    closed: uint8

  RingBuffer = object
    hdr: ptr RingHeader
    data: ptr UncheckedArray[byte]
    capacity: uint64

proc initRing(base: pointer, capacity: uint64): RingBuffer =
  result.hdr = cast[ptr RingHeader](base)
  result.data = cast[ptr UncheckedArray[byte]](cast[uint](base) + sizeof(RingHeader).uint)
  result.capacity = capacity

proc setup(r: var RingBuffer, capacity: uint64) =
  r.hdr.writeIdx.store(0)
  r.hdr.readIdx.store(0)
  r.hdr.capacity = capacity
  r.hdr.closed = 0

proc writeBytes(r: var RingBuffer, pos: uint64, data: pointer, n: int) =
  let p = pos mod r.capacity
  if p.int + n <= r.capacity.int:
    copyMem(addr r.data[p.int], data, n)
  else:
    let first = r.capacity.int - p.int
    copyMem(addr r.data[p.int], data, first)
    copyMem(addr r.data[0], cast[pointer](cast[uint](data) + first.uint), n - first)

proc readBytesOut(r: var RingBuffer, pos: uint64, outp: pointer, n: int) =
  let p = pos mod r.capacity
  if p.int + n <= r.capacity.int:
    copyMem(outp, addr r.data[p.int], n)
  else:
    let first = r.capacity.int - p.int
    copyMem(outp, addr r.data[p.int], first)
    copyMem(cast[pointer](cast[uint](outp) + first.uint), addr r.data[0], n - first)

proc tryWrite(r: var RingBuffer, payload: pointer, length: uint32): bool =
  let need = 4'u64 + length.uint64
  let w = r.hdr.writeIdx.load(moRelaxed)
  let rIdx = r.hdr.readIdx.load(moAcquire)
  let used = (w - rIdx) mod (2 * r.capacity)
  if r.capacity - used < need: return false
  var lenBuf = length
  r.writeBytes(w, addr lenBuf, 4)
  r.writeBytes(w + 4, payload, length.int)
  r.hdr.writeIdx.store((w + need) mod (2 * r.capacity), moRelease)
  true

proc tryRead(r: var RingBuffer, outp: pointer, outCap: int): int =
  let w = r.hdr.writeIdx.load(moAcquire)
  let rIdx = r.hdr.readIdx.load(moRelaxed)
  if w == rIdx: return -1
  var length: uint32
  r.readBytesOut(rIdx, addr length, 4)
  r.readBytesOut(rIdx + 4, outp, length.int)
  r.hdr.readIdx.store((rIdx + 4 + length.uint64) mod (2 * r.capacity), moRelease)
  length.int

proc isClosed(r: RingBuffer): bool = r.hdr.closed == 1

proc main() =
  let totalSize = sizeof(RingHeader) + Capacity
  let name = "/nim_bench_ring"
  let fd = shm_open(name.cstring, O_CREAT or O_RDWR, 0o666)
  discard ftruncate(fd, totalSize.Off)
  let base = mmap(nil, totalSize, PROT_READ or PROT_WRITE, MAP_SHARED, fd, 0)

  var ring = initRing(base, Capacity.uint64)
  ring.setup(Capacity.uint64)

  const n = 500_000

  let pid = fork()
  if pid == 0:
    # child = producer
    let fd2 = shm_open(name.cstring, O_RDWR, 0o666)
    let base2 = mmap(nil, totalSize, PROT_READ or PROT_WRITE, MAP_SHARED, fd2, 0)
    var prod = initRing(base2, Capacity.uint64)
    for i in 0 ..< n:
      var payload = i.int64
      while not prod.tryWrite(addr payload, sizeof(payload).uint32):
        discard sched_yield()
    prod.hdr.closed = 1
    quit(0)

  # parent = consumer
  var outBuf: array[64, byte]
  var received = 0
  let t0 = getMonoTime()
  while received < n:
    let got = ring.tryRead(addr outBuf[0], outBuf.len)
    if got < 0:
      if ring.isClosed: break
      discard sched_yield()
      continue
    doAssert got == sizeof(int64)
    received.inc

  let t1 = getMonoTime()
  var status: cint
  discard waitpid(pid, status, 0)

  let secs = (t1 - t0).inNanoseconds.float / 1e9
  echo &"=== Nim SPSC ring buffer, cross-process (fork), N={n} ==="
  echo &"  received {received} messages in {secs:.4f}s  ({received.float/secs:.0f} msg/s)"

  discard munmap(base, totalSize)
  discard shm_unlink(name.cstring)

main()

File: cpp/bench_reference/bench_ringbuffer.cpp
// bench_ringbuffer.cpp
// Cross-process SPSC ring buffer over POSIX shared memory, same
// design as shared_memory_ipc.py (doubled-modulus indices, spin/yield
// backoff), benchmarked the same way as test_shm.py: fork a producer,
// have the parent consume, time until all messages are received.
#include <atomic>
#include <cstdint>
#include <cstring>
#include <cstdio>
#include <chrono>
#include <thread>
#include <fcntl.h>
#include <sys/mman.h>
#include <sys/wait.h>
#include <unistd.h>

struct RingHeader {
    std::atomic<uint64_t> write_idx;
    std::atomic<uint64_t> read_idx;
    uint64_t capacity;
    uint8_t closed;
};

class RingBuffer {
public:
    RingBuffer(uint8_t* base, uint64_t capacity)
        : hdr_(reinterpret_cast<RingHeader*>(base)),
          data_(base + sizeof(RingHeader)),
          capacity_(capacity) {}

    void init() {
        hdr_->write_idx.store(0);
        hdr_->read_idx.store(0);
        hdr_->capacity = capacity_;
        hdr_->closed = 0;
    }

    bool try_write(const uint8_t* payload, uint32_t len) {
        uint64_t need = 4 + len;
        uint64_t w = hdr_->write_idx.load(std::memory_order_relaxed);
        uint64_t r = hdr_->read_idx.load(std::memory_order_acquire);
        uint64_t used = (w - r) % (2 * capacity_);
        if (capacity_ - used < need) return false;
        write_bytes(w, reinterpret_cast<const uint8_t*>(&len), 4);
        write_bytes(w + 4, payload, len);
        hdr_->write_idx.store((w + need) % (2 * capacity_), std::memory_order_release);
        return true;
    }

    // returns length written into out, or -1 if empty
    int try_read(uint8_t* out, uint32_t out_cap) {
        uint64_t w = hdr_->write_idx.load(std::memory_order_acquire);
        uint64_t r = hdr_->read_idx.load(std::memory_order_relaxed);
        if (w == r) return -1;
        uint32_t len;
        read_bytes(r, reinterpret_cast<uint8_t*>(&len), 4);
        read_bytes(r + 4, out, len);
        hdr_->read_idx.store((r + 4 + len) % (2 * capacity_), std::memory_order_release);
        return (int)len;
    }

    void mark_closed() { hdr_->closed = 1; }
    bool is_closed() const { return hdr_->closed == 1; }

private:
    void write_bytes(uint64_t pos, const uint8_t* data, uint32_t n) {
        pos %= capacity_;
        if (pos + n <= capacity_) {
            std::memcpy(data_ + pos, data, n);
        } else {
            uint32_t first = capacity_ - pos;
            std::memcpy(data_ + pos, data, first);
            std::memcpy(data_, data + first, n - first);
        }
    }
    void read_bytes(uint64_t pos, uint8_t* out, uint32_t n) {
        pos %= capacity_;
        if (pos + n <= capacity_) {
            std::memcpy(out, data_ + pos, n);
        } else {
            uint32_t first = capacity_ - pos;
            std::memcpy(out, data_ + pos, first);
            std::memcpy(out + first, data_, n - first);
        }
    }

    RingHeader* hdr_;
    uint8_t* data_;
    uint64_t capacity_;
};

int main() {
    const uint64_t CAPACITY = 1 << 20;
    const size_t TOTAL = sizeof(RingHeader) + CAPACITY;
    const int N = 500000;  // messages, matching the scale of the Python cross-process test but larger for a stable measurement

    int fd = shm_open("/cpp_bench_ring", O_CREAT | O_RDWR, 0666);
    ftruncate(fd, TOTAL);
    uint8_t* base = (uint8_t*)mmap(nullptr, TOTAL, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);

    RingBuffer ring(base, CAPACITY);
    ring.init();

    pid_t pid = fork();
    if (pid == 0) {
        // child = producer
        int fd2 = shm_open("/cpp_bench_ring", O_RDWR, 0666);
        uint8_t* base2 = (uint8_t*)mmap(nullptr, TOTAL, PROT_READ | PROT_WRITE, MAP_SHARED, fd2, 0);
        RingBuffer prod(base2, CAPACITY);
        int64_t payload;
        for (int i = 0; i < N; i++) {
            payload = i;
            while (!prod.try_write(reinterpret_cast<uint8_t*>(&payload), sizeof(payload))) {
                std::this_thread::yield();
            }
        }
        prod.mark_closed();
        _exit(0);
    }

    // parent = consumer
    uint8_t out[64];
    int received = 0;
    auto t0 = std::chrono::high_resolution_clock::now();
    while (received < N) {
        int n = ring.try_read(out, sizeof(out));
        if (n < 0) {
            if (ring.is_closed()) break;
            std::this_thread::yield();
            continue;
        }
        received++;
    }
    auto t1 = std::chrono::high_resolution_clock::now();
    waitpid(pid, nullptr, 0);

    double secs = std::chrono::duration<double>(t1 - t0).count();
    printf("=== C++ SPSC ring buffer, cross-process (fork), N=%d ===\n", N);
    printf("  received %d messages in %.4fs  (%.0f msg/s)\n", received, secs, received / secs);

    munmap(base, TOTAL);
    shm_unlink("/cpp_bench_ring");
    return 0;
}
